File: src/main/java/org/example/binarysearch/Search2DMatrix.java
package org.example.binarysearch;

public class Search2DMatrix {
    public boolean searchMatrix(int[][] matrix, int target) {
        for (int[] row : matrix) {
            if (target <= row[row.length - 1])
                return binarySearch(row, target);
        }

        return false;
    }

    private boolean binarySearch(int[] nums, int target) {
        int left = 0, right = nums.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (target < nums[mid]) {
                right = mid - 1;
            } else if (target > nums[mid]) {
                left = mid + 1;
            } else {
                return true;
            }
        }

        return false;
    }
}

File: src/main/java/org/example/binarysearch/Search2DMatrix.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        ROWS, COLS = len(matrix), len(matrix[0])

        def search(nums):
            l, r = 0, len(nums) - 1
            while l <= r:
                mid = l + (r - l) // 2
                if target < nums[mid]:
                    r = mid - 1
                elif target > nums[mid]:
                    l = mid + 1
                else:
                    return True
            return False

        for r in range(ROWS):
            if target <= matrix[r][COLS - 1]:
                return search(matrix[r])


obj = Solution()
#matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]]
#target = 3
matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]]
target = 13
#matrix = [[1], [3]]
#target = 3
print(obj.searchMatrix(matrix, target))
